src/aro/Driver.zig

```diff
@@ -892,3 +916,214 @@ fn exitWithCleanup(d: *Driver, code: u8) noreturn {
+    // The Darwin/MachO default to use PIC does not apply when using -static.
+    if (target.ofmt == .macho and d.static) {
+        pic, pie = .{ false, false };
+    }
+    var is_piclevel_two = pic;
+
+    const kernel_or_kext: bool = d.mkernel or d.apple_kext;
+
+    // Android-specific defaults for PIC/PIE
+    if (target.isAndroid()) {
+        switch (target.cpu.arch) {
+            .arm,
+            .armeb,
+            .thumb,
+            .thumbeb,
+            .aarch64,
+            .mips,
+            .mipsel,
+            .mips64,
+            .mips64el,
+            => pic = true, // "-fpic"
+
+            .x86, .x86_64 => {
+                pic = true; // "-fPIC"
+                is_piclevel_two = true;
+            },
+            else => {},
+        }
+    }
+
+    // OHOS-specific defaults for PIC/PIE
+    if (target.abi == .ohos and target.cpu.arch == .aarch64)
+        pic = true;
+
+    // OpenBSD-specific defaults for PIE
+    if (target.os.tag == .openbsd) {
+        switch (target.cpu.arch) {
+            .arm, .aarch64, .mips64, .mips64el, .x86, .x86_64 => is_piclevel_two = false, // "-fpie"
+            .powerpc, .sparc64 => is_piclevel_two = true, // "-fPIE"
+            else => {},
+        }
+    }
+
+    // The last argument relating to either PIC or PIE wins, and no
+    // other argument is used. If the last argument is any flavor of the
+    // '-fno-...' arguments, both PIC and PIE are disabled. Any PIE
+    // option implicitly enables PIC at the same level.
+    if (target.os.tag == .windows and
+        !target_util.isCygwinMinGW(target) and
+        // d.lastpic_pos != null and
+        StaticStringSet.initComptime(.{ .{"-fPIC"}, .{"-fpic"}, .{"-fpie"}, .{"-fPIE"} }).has(lastpic))
+    {
+        try d.unsupportedOptionForTarget(target, lastpic);
+        if (target.cpu.arch == .x86_64)
+            return .{ .two, false };
+        return .{ .none, false };
+    }
+
+    // Check whether the tool chain trumps the PIC-ness decision. If the PIC-ness
+    // is forced, then neither PIC nor PIE flags will have no effect.
+    const forced = switch (target_util.isPICDefaultForced(target)) {
+        .yes => true,
+        .no => false,
+        .depends_on_linker => if (is_bfd_linker) target.cpu.arch == .x86_64 else target.cpu.arch == .aarch64 or target.cpu.arch == .x86_64,
+    };
+    if (!forced) {
+        if (StaticStringSet.initComptime(.{ .{"-fPIC"}, .{"-fpic"}, .{"-fpie"}, .{"-fPIE"} }).has(lastpic)) {
+            pie = StaticStringSet.initComptime(.{ .{"-fpie"}, .{"-fPIE"} }).has(lastpic);
+            pic = pie or StaticStringSet.initComptime(.{ .{"-fpic"}, .{"-fPIC"} }).has(lastpic);
+            is_piclevel_two = StaticStringSet.initComptime(.{ .{"-fPIE"}, .{"-fPIC"} }).has(lastpic);
+        } else {
+            pic, pie = .{ false, false };
+            if (target_util.isPS(target)) {
+                const model_arg = getLastArgValue(args, "-mcmodel=");
```

Comment:
Could you move the remaining args as well? The `getLastArg*` would not exist ideally.

Reply:
Yeah! I will clean up the rest tomorrow.